src/python.cpp

```diff
@@ -2207,3 +2207,98 @@ SEXP py_eval_impl(const std::string& code, bool convert = true) {
   return result;
 
 }
+
+// [[Rcpp::export]]
+SEXP py_convert_pandas_series(PyObjectRef series) {
+  
+  // extract dtype
+  PyObjectPtr dtype(PyObject_GetAttrString(series, "dtype"));
+  PyObjectPtr name(PyObject_GetAttrString(dtype, "name"));
+  
+  // as well as values
+  PyObjectPtr values(PyObject_GetAttrString(series, "values"));
+  
+  RObject R_obj;
+  
+  // special treatment for pd.Categorical
+  if (as_std_string(name) == "category") {
+    
+    // get actual values and convert to R
+    PyObjectPtr cat(PyObject_GetAttrString(series, "cat"));
+    PyObjectPtr codes(PyObject_GetAttrString(cat, "codes"));
+    PyObjectPtr code_values(PyObject_GetAttrString(codes, "values"));
+    RObject R_values = py_to_r(code_values, true);
+    
+    // get levels and convert to R
+    PyObjectPtr categories(PyObject_GetAttrString(dtype, "categories"));
+    PyObjectPtr category_values(PyObject_GetAttrString(categories, "values"));
+    RObject R_levels = py_to_r(category_values, true);
+    
+    // get "ordered" attribute
+    PyObjectPtr ordered(PyObject_GetAttrString(dtype, "ordered"));
+    //RObject ordered = py_to_r(ordered_, true);
+    
+    // populate integer vector to hold factor values
+    int* codes_int = INTEGER(R_values);
+    int n = Rf_xlength(R_values);
+    IntegerVector factor(n);
+    // values need to start at 1
+    for (int i = 0; i < n; ++i) {
+      factor[i] = codes_int[i] + 1;
+    }
+    
+    // populate character vector to hold levels
+    CharacterVector factor_levels(R_levels);
+    factor_levels.attr("dim") = R_NilValue;
+    
+    factor.attr("class") = "factor";
+    factor.attr("levels") = factor_levels;
+    if (PyObject_IsTrue(ordered)) factor.attr("ordered") = true;
+    
+    R_obj = factor;
+    
+    // default case
+  } else {
+    
+    R_obj = py_to_r(values, series.convert());
+    
+  }
+  
+  return R_obj;
+  
+}
+
+// [[Rcpp::export]]
+SEXP py_convert_pandas_df(PyObjectRef df) {
+  
+  // pd.DataFrame.items() returns an Iterator over (column name, Series) pairs
+  PyObjectPtr items(PyObject_CallMethod(df, "items", NULL));
+  if (! (PyObject_HasAttrString(items, "__next__") || PyObject_HasAttrString(items, "next")))
+    stop("Cannot iterate over object");
+
+  std::vector<RObject> list;
+  
+  while (true) {
+
+    // get next tuple
+    PyObjectPtr tuple(PyIter_Next(items));
+    if (tuple.is_null()) {
+      if (PyErr_Occurred())
+        stop(py_fetch_error());
+      else
+        break;
+    }
+    
+    // access Series in slot 1
+    PyObjectPtr series(PySequence_GetItem(tuple, 1));
+    // delegate to py_convert_pandas_series
+    PyObjectRef series_ref(series, df.convert());
```

Comment:
Now that I think about it, because this PyObjectRef will now "own" the previously-created `series` object, I believe you will have to use `series.detach()` here. See e.g.

https://github.com/rstudio/reticulate/blob/86acdf16f9208cf4952b1a312bccdc91f7a25bdf/src/python.cpp#L2202-L2205

Reply:
Hmm, but we aren't actually returning this reference back to R -- we're only using it temporarily while we convert that object back into an R object...

Reply:
Actually yes -- I believe this is indeed the issue. PyObjectRef will also call Py_DecRef once the associated R object is cleaned up by the garbage collector; e.g. by

https://github.com/rstudio/reticulate/blob/86acdf16f9208cf4952b1a312bccdc91f7a25bdf/src/reticulate_types.h#L9-L13

So now we have two calls to `Py_DecRef()`, but only a single `Py_IncRef()` (implicitly through the construction of the `PyObjectPtr`). So `series.detach()` should indeed be the fix.

Reply:
That was it :-) :-)
thanks!!

(I hadn't really understood the `detach` so far, but now I have an impressive example... :-))

((yes and I had seen `py_ref` and checked if it did something that might make a difference in that respect, but as I saw it didn't and then it seemed like unnecessary overhead given the object was just a temporary construction)))